pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,76 +196,406 @@ public void shutdown() {
   public void markDataResourceOnline(String tableName, ExternalView externalView,
       List<InstanceConfig> instanceConfigList) {
     if (externalView == null) {
+      // It is possible for us to get a request to serve a table for which there is no ideal state. In this case, just
+      // keep a bogus last seen external view version to force a rebuild the next time we see an external view
+      _lastKnownExternalViewVersionMap.put(tableName, INVALID_EXTERNAL_VIEW_VERSION);
       return;
     }
 
-    long startTime = System.currentTimeMillis();
+    // Build the routing table
+    buildRoutingTable(tableName, externalView, instanceConfigList);
+  }
+
+  private boolean isRoutingTableRebuildRequired(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigs) {
+    // In unit tests, always rebuild the routing table
+    if (_helixManager == null) {
+      return true;
+    }
 
+    // Do we know about this table?
+    if (!_lastKnownExternalViewVersionMap.containsKey(tableName)) {
+      LOGGER.info("Routing table for table {} requires rebuild due to it being newly added", tableName);
+      return true;
+    }
+
+    // Check if the znode version changed
     int externalViewRecordVersion = externalView.getRecord().getVersion();
-    if (_routingTableLastKnownZkVersionMap.containsKey(tableName)) {
-      long lastKnownZkVersion = _routingTableLastKnownZkVersionMap.get(tableName);
-      if (externalViewRecordVersion == lastKnownZkVersion) {
-        LOGGER.info(
-            "No change on routing table version (current version {}, last known version {}), do nothing for table {}",
-            externalViewRecordVersion, lastKnownZkVersion, tableName);
-        return;
+    int lastKnownExternalViewVersion = _lastKnownExternalViewVersionMap.get(tableName);
+
+    if (externalViewRecordVersion != lastKnownExternalViewVersion || lastKnownExternalViewVersion == INVALID_EXTERNAL_VIEW_VERSION) {
```

Comment:
In the unlikely event that the external view znode version is equal to `Integer.MIN_VALUE`, we'll always rebuild the routing table in this case. It shouldn't happen, but if it does, then we won't lose that routing table update.